src/compiler/checker.ts

```diff
@@ -134,6 +134,8 @@ namespace ts {
         const anySignature = createSignature(undefined, undefined, emptyArray, anyType, 0, /*hasRestParameter*/ false, /*hasStringLiterals*/ false);
         const unknownSignature = createSignature(undefined, undefined, emptyArray, unknownType, 0, /*hasRestParameter*/ false, /*hasStringLiterals*/ false);
 
+        const enumNumberIndexInfo = createIndexInfo(stringType, /*isReadonly*/ false);

```

Comment:
That's right. It's bit of a puzzle ensure backwards compatibility, so there are actually two forms of read-only-ness in the new implementation:

* `readonly` properties, `readonly` index signatures, and `get` only accessors are *always* considered read-only, in both assignments and type compatibility checks.
* `const` variables, functions, classes, namespaces, enums, and enum members are considered read-only in assignments, *but not* in type compatibility checks. This ensures backwards compatibility. For example, it is reasonably common to implement an interface contract using a module or namespace, and strict read-only-ness would break that pattern.